clients/src/main/java/org/apache/kafka/common/record/LegacyRecordBatch.java

```diff
@@ -0,0 +1,454 @@
+            @Override
+            protected Record makeNext() {
+                if (iterator.hasNext())
+                    return iterator.next();
+                return allDone();
+            }
+        };
+    }
+
+    public static void writeHeader(ByteBuffer buffer, long offset, int size) {
+        buffer.putLong(offset);
+        buffer.putInt(size);
+    }
+
+    public static void writeHeader(DataOutputStream out, long offset, int size) throws IOException {
+        out.writeLong(offset);
+        out.writeInt(size);
+    }
+
+    private static final class DataLogInputStream implements LogInputStream<LegacyRecordBatch> {
+        private final DataInputStream stream;
+        protected final int maxMessageSize;
+
+        DataLogInputStream(DataInputStream stream, int maxMessageSize) {
+            this.stream = stream;
+            this.maxMessageSize = maxMessageSize;
+        }
+
+        public LegacyRecordBatch nextBatch() throws IOException {
+            try {
+                long offset = stream.readLong();
+                int size = stream.readInt();
+                if (size < LegacyRecord.RECORD_OVERHEAD_V0)
+                    throw new CorruptRecordException(String.format("Record size is less than the minimum record overhead (%d)", LegacyRecord.RECORD_OVERHEAD_V0));
+                if (size > maxMessageSize)
+                    throw new CorruptRecordException(String.format("Record size exceeds the largest allowable message size (%d).", maxMessageSize));
+
+                byte[] recordBuffer = new byte[size];
+                stream.readFully(recordBuffer, 0, size);
+                ByteBuffer buf = ByteBuffer.wrap(recordBuffer);
+                return new BasicLegacyRecordBatch(offset, new LegacyRecord(buf));
+            } catch (EOFException e) {
+                return null;
+            }
+        }
+    }
+
+    private static class DeepRecordsIterator extends AbstractIterator<LegacyRecordBatch> {
+        private final ArrayDeque<LegacyRecordBatch> logEntries;
+        private final long absoluteBaseOffset;
+        private final byte wrapperMagic;
+
+        private DeepRecordsIterator(LegacyRecordBatch wrapperEntry, boolean ensureMatchingMagic, int maxMessageSize) {
+            LegacyRecord wrapperRecord = wrapperEntry.record();
+            this.wrapperMagic = wrapperRecord.magic();
+
+            CompressionType compressionType = wrapperRecord.compressionType();
+            ByteBuffer wrapperValue = wrapperRecord.value();
+            if (wrapperValue == null)
+                throw new InvalidRecordException("Found invalid compressed record set with null value (magic = " +
+                        wrapperMagic + ")");
+
+            DataInputStream stream = new DataInputStream(compressionType.wrapForInput(
+                    new ByteBufferInputStream(wrapperValue), wrapperRecord.magic()));
+            LogInputStream<LegacyRecordBatch> logStream = new DataLogInputStream(stream, maxMessageSize);
```

Comment:
I'll clarify the comment. What I was trying to say is that `LogInputStream` only handles one level of iteration (it does not descend into compressed payloads).